Add "dynamic elevation range per band" mode for rasters

nyalldawson opened the pull request:
This accompanies https://github.com/qgis/QGIS/pull/56889, and exposes a  new mode where the elevation range for raster bands are calculated using a pair of qgis expressions (for the lower and upper elevation range).

![image](https://github.com/qgis/QGIS/assets/1829991/7af4ed05-2eb6-40de-a6e4-439b8670cd1a)

In this mode, the user can specify a QGIS expression for the lower and upper value corresponding to raster bands, using variables like @band, @band_name and @band_description.
    
E.g
    
    @band * 100
    
Can be used when each band represents a 100 m vertical slice of data. (You can mix these with all the other standard qgis global/project/layer variables and expression functions)
    
The expression will be evaluated when required to determine the actual elevation range corresponding to each band.
    
This differs from the other "Fixed Elevation Range Per Band" mode in that "Fixed Elevation Range Per Band" requires users  to manually enter an elevation for each band separately, and these values are then treated as constants. That mode works    best for rasters with non-regular steps in the band elevation values, while this new mode is better for regular band elevation steps.

nyalldawson committed:
Add "dynamic elevation range per band" mode for rasters

In this mode, the user can specify a QGIS expression for the
lower and upper value corresponding to raster bands, using
variables like @band, @band_name and @band_description.

E.g

    @band * 100

Can be used when each band represents a 100 m vertical slice
of data.

The expression will be evaluated when required to determine
the actual elevation range corresponding to each band.

This differs from the existing "Fixed Elevation Range Per Band"
mode in that "Fixed Elevation Range Per Band" requires users
to manually enter an elevation for each band separately,
and these values are then treated as constants. That mode works
best for rasters with non-regular steps in the band
elevation values, while this new mode is better for regular
band elevation steps

nyalldawson committed:
Fix filtering raster identify results by elevation filter

wonder-sk reviewed (approved):
The code looks good.

One thought about GUI: if I understand correctly, in the dynamic mode, user does not really see if the expressions are correctly evaluated for bands, so it can be a bit of a guessing game (or requiring to switch to "fixed" mode first). How about both fixed and dynamic mode were condensed into a single combo box entry, and there would be a toggle between dynamic mode (with read-only table) and fixed mode (with editable table) ?

nyalldawson committed:
Add API for custom preview generators in QgsExpressionBuilderWidget

In this mode, the widget will call a callback function to generate
a new QgsExpressionContext as the previewed object changes. This
can be used to provide custom preview values for different objects
(i.e. for objects which aren't vector layer features), such as raster
bands or other custom objects.

nyalldawson committed:
Allow picking band to preview expression result for

Populates the expression builders with the raster bands, so that
the preview control allows users to scroll through bands and
see the expression results for each.

nyalldawson committed:
Show a read only table of band ranges in dialog when in dynamic range
per band mode

nyalldawson commented:
@wonder-sk 
> One thought about GUI: if I understand correctly, in the dynamic mode, user does not really see if the expressions are correctly evaluated for bands, so it can be a bit of a guessing game (or requiring to switch to "fixed" mode first). How about both fixed and dynamic mode were condensed into a single combo box entry, and there would be a toggle between dynamic mode (with read-only table) and fixed mode (with editable table) ?

Ok, fair enough! I've gone with a slight variation on this, as I really want to keep these two options as distinct choices in the combo box.

Now, in the dynamic mode there's a read only table showing the calculated ranges for the bands:

![Peek 2024-03-21 15-25](https://github.com/qgis/QGIS/assets/1829991/4e1ea83c-578b-48f9-b908-8ab3179950e0)

I've also made it so the expression preview widget in the builder now does something, and allows users to step through the bands to preview the results in the expression builder itself:

![Peek 2024-03-21 15-22](https://github.com/qgis/QGIS/assets/1829991/4331c531-b3fe-492f-823c-ead6e4743b0d)

nyalldawson committed:
Fix warning

wonder-sk commented:
@nyalldawson Great - thanks for the updates!
